Initializing variables and libraries. Importing data from CSV file

In [5]:
import numpy as np
import tensorflow as tf
import pandas as pd

path="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_train_features.csv"
data=pd.read_csv(path)
pathTest="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_test_features.csv"
dataTest=pd.read_csv(pathTest)
XData=data.iloc[:,0:30].values
XDataTest=dataTest.iloc[:,0:30].values

Normalizing data and creating labels for overall survival.

In [6]:
for i in range(30):
    XData[:,i]=(XData[:,i]-np.mean(XData[:,i]))/(np.std(XData[:,i]))
    XDataTest[:,i]=(XDataTest[:,i]-np.mean(XDataTest[:,i]))/(np.std(XDataTest[:,i]))
bias=tf.Variable(tf.zeros(1,dtype=tf.float64))
YData=np.transpose(np.matrix(np.float64(data.iloc[:,30].values)))
YDataTest=np.transpose(np.matrix(np.float64(dataTest.iloc[:,30].values)))
for i in range(98):
    if YData[i,0]>=300:
        YData[i,0]=1
    else:
        YData[i,0]=0

for i in range(len(YDataTest)):   
    if YDataTest[i,0]>=300:
        YDataTest[i,0]=1
    else:
        YDataTest[i,0]=0

Tensorflow placeholders and variables defined along with gradient descent optimization call.

In [9]:
X=tf.placeholder(tf.float64,shape=(None,30))
Y=tf.placeholder(tf.float64,shape=(98,1))
y_ = tf.placeholder(tf.float64, [33,1])
weights=tf.Variable(tf.ones([30,1],tf.float64))
yPred=tf.nn.sigmoid(-(tf.matmul(X,weights)+bias))
loss=-tf.reduce_sum(Y*tf.log(yPred)+(1-Y)*tf.log(1-yPred))
optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.0005).minimize(loss)

Tensorflow session with 1000 iterations. Accuracy is output using the same model on test data.

In [21]:
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for i in range(1000):
        sess.run([optimizer], feed_dict={X:XData, Y:YData})
    yPred=tf.cast(tf.greater_equal(yPred,tf.constant(0.5,dtype=tf.float64)),tf.float64)
    correct_prediction = tf.equal(yPred,y_)
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float64))
    print(sess.run(accuracy, feed_dict={X: XDataTest, y_: YDataTest}))
    prod=tf.multiply(yPred,y_)
    summation=yPred+y_
    ab=tf.cast(tf.size(y_),tf.float64)-tf.reduce_sum(y_)
    specificity=tf.reduce_sum(tf.cast(tf.equal(yPred,tf.constant(0.0,tf.float64)),tf.float64))/ab
    truepos=tf.reduce_sum(prod)
    sensitivity=truepos/tf.reduce_sum(y_)
    print(sess.run([sensitivity,specificity], feed_dict={X: XDataTest, y_: YDataTest}))

0.8181818181818182
[0.8947368421052632, 0.8571428571428571]


81.81% accuracy is achieved on test data. Sensitivity value is 89.473% and specificity is 85.714%

In [1]:
# -*- coding: utf-8 -*-
"""
Created on Fri Mar  2 21:54:25 2018

@author: srini
"""
import numpy as np
import tensorflow as tf
import pandas as pd

path="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_train_features.csv"
data=pd.read_csv(path)
pathTest="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_test_features.csv"
dataTest=pd.read_csv(pathTest)
XData=data.iloc[:,0:30].values
XDataTest=dataTest.iloc[:,0:30].values
for i in range(30):
    XData[:,i]=(XData[:,i]-np.mean(XData[:,i]))/(np.std(XData[:,i]))
    XDataTest[:,i]=(XDataTest[:,i]-np.mean(XDataTest[:,i]))/(np.std(XDataTest[:,i]))
bias=tf.Variable(tf.zeros(1,dtype=tf.float64))
#XData=np.column_stack((XData,bias))
YData=np.transpose(np.matrix(np.float64(data.iloc[:,30].values)))
YDataTest=np.transpose(np.matrix(np.float64(dataTest.iloc[:,30].values)))
for i in range(98):
    if YData[i,0]>=300:
        YData[i,0]=1
    #elif (YData[i,0]>=300) & (YData[i,0]<450):
    #    YData[i,0]=1
    else:
        YData[i,0]=0

for i in range(len(YDataTest)):   
    if YDataTest[i,0]>=300:
        YDataTest[i,0]=1
    #elif (YDataTest[i,0]>=300) & (YDataTest[i,0]<450:
    #    YDataTest[i,0]=1
    else:
        YDataTest[i,0]=0
X=tf.placeholder(tf.float64,shape=(None,30))
Y=tf.placeholder(tf.float64,shape=(98,1))
y_ = tf.placeholder(tf.float64, [33,1])
weights=tf.Variable(tf.ones([30,1],tf.float64))
yPred=tf.nn.sigmoid(-(tf.matmul(X,weights)+bias))
loss=-tf.reduce_sum(Y*tf.log(yPred)+(1-Y)*tf.log(1-yPred))
optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.05).minimize(loss)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for i in range(1000):
        sess.run([optimizer], feed_dict={X:XData, Y:YData})
    #weights, loss, bias=sess.run([weights,loss,bias], feed_dict={X:XData, Y:YData})
    #yPred=1/(1+tf.exp(-tf.matmul(XData,weights)))
    #yPred=tf.cast(tf.greater_equal(yPred,tf.constant(0.5,dtype=tf.float64)),tf.float64)
    #pred, truth=sess.run([yPred,Y], feed_dict={X:XData, Y:YData})
    #print(np.mean(pred==truth))
#    yPredTest=1/(1+np.exp(-(np.dot(XDataTest,weights)+bias)))
#    yPredTest=np.cast['f'](yPredTest>0.5)
#    print(np.mean(yPredTest==YDataTest))
    correct_prediction = tf.equal(tf.argmax(yPred,1), tf.argmax(y_,1))
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float64))
    print(sess.run(accuracy, feed_dict={X: XDataTest, y_: YDataTest}))


1.0


In [22]:
import numpy as np
import tensorflow as tf
import pandas as pd

path="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_train_features.csv"
data=pd.read_csv(path)
pathTest="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_test_features.csv"
dataTest=pd.read_csv(pathTest)
XData=data.iloc[:,0:30].values
XDataTest=dataTest.iloc[:,0:30].values
for i in range(30):
    XData[:,i]=(XData[:,i]-np.mean(XData[:,i]))/(np.std(XData[:,i]))
    XDataTest[:,i]=(XDataTest[:,i]-np.mean(XDataTest[:,i]))/(np.std(XDataTest[:,i]))
bias=tf.Variable(tf.zeros(1,dtype=tf.float64))
YData=np.transpose(np.matrix(np.float64(data.iloc[:,30].values)))
YDataTest=np.transpose(np.matrix(np.float64(dataTest.iloc[:,30].values)))
for i in range(98):
    if YData[i,0]>=300:
        YData[i,0]=1
    else:
        YData[i,0]=0

for i in range(len(YDataTest)):   
    if YDataTest[i,0]>=300:
        YDataTest[i,0]=1
    else:
        YDataTest[i,0]=0
X=tf.placeholder(tf.float64,shape=(None,30))
Y=tf.placeholder(tf.float64,shape=(98,1))
y_ = tf.placeholder(tf.float64, [33,1])
weights=tf.Variable(tf.ones([30,1],tf.float64))
yPred=tf.nn.sigmoid(-(tf.matmul(X,weights)+bias))
loss=-tf.reduce_sum(Y*tf.log(yPred)+(1-Y)*tf.log(1-yPred))
optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.0005).minimize(loss)
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for i in range(1000):
        sess.run([optimizer], feed_dict={X:XData, Y:YData})
    yPred=tf.cast(tf.greater_equal(yPred,tf.constant(0.5,dtype=tf.float64)),tf.float64)
    correct_prediction = tf.equal(yPred,y_)
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float64))
    print(sess.run(accuracy, feed_dict={X: XDataTest, y_: YDataTest}))
    prod=tf.multiply(yPred,y_)
    summation=yPred+y_
    ab=tf.cast(tf.size(y_),tf.float64)-tf.reduce_sum(y_)
    specificity=tf.reduce_sum(tf.cast(tf.equal(yPred,tf.constant(0.0,tf.float64)),tf.float64))/ab
    truepos=tf.reduce_sum(prod)
    sensitivity=truepos/tf.reduce_sum(y_)
    print(sess.run([sensitivity,specificity], feed_dict={X: XDataTest, y_: YDataTest}))

0.8181818181818182
[0.8947368421052632, 0.8571428571428571]


Multinomial logistic Regression

In [39]:
# -*- coding: utf-8 -*-
"""
Created on Fri Mar  2 21:54:25 2018

@author: srini
"""
import numpy as np
import tensorflow as tf
import pandas as pd

path="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_train_features.csv"
data=pd.read_csv(path)
pathTest="C:\\Users\\srini\\OneDrive\\Academics\\ML courses\\ID 5030\\Assignments\\Assignment 3\\30_test_features.csv"
dataTest=pd.read_csv(pathTest)
XData=data.iloc[:,0:30].values
XDataTest=dataTest.iloc[:,0:30].values
for i in range(30):
    XData[:,i]=(XData[:,i]-np.mean(XData[:,i]))/(np.std(XData[:,i]))
    XDataTest[:,i]=(XDataTest[:,i]-np.mean(XDataTest[:,i]))/(np.std(XDataTest[:,i]))
bias=tf.Variable(tf.zeros(3,dtype=tf.float64))
extra=np.zeros([98,1])
extraTest=np.zeros([33,1])
YData=np.column_stack((np.transpose(np.matrix(np.float64(data.iloc[:,30].values))),extra,extra))
YDataTest=np.column_stack((np.transpose(np.matrix(np.float64(dataTest.iloc[:,30].values))),extraTest,extraTest))
for i in range(98):
    if YData[i,0]>=450:
        YData[i,0]=1
        YData[i,1]=0
        YData[i,2]=0
    elif (YData[i,0]>=300) & (YData[i,0]<450):
        YData[i,0]=0
        YData[i,1]=1
        YData[i,2]=0
    else:
        YData[i,0]=0
        YData[i,1]=0
        YData[i,2]=1

for i in range(len(YDataTest)):   
    if YDataTest[i,0]>=450:
        YDataTest[i,0]=1
        YDataTest[i,1]=0
        YDataTest[i,2]=0
    elif (YDataTest[i,0]>=300) & (YDataTest[i,0]<450):
        YDataTest[i,0]=0
        YDataTest[i,1]=1
        YDataTest[i,2]=0
    else:
        YDataTest[i,0]=0
        YDataTest[i,1]=0
        YDataTest[i,2]=1

X=tf.placeholder(tf.float64,shape=(None,30))
Y=tf.placeholder(tf.float64,shape=(None,3))
weights=tf.Variable(tf.ones([30,3],tf.float64))
yPred=tf.nn.softmax(tf.matmul(X,weights)+bias)  
loss=-tf.reduce_sum(Y*tf.log(yPred)+(1-Y)*tf.log(1-yPred))
optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.005).minimize(loss)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for i in range(1000):
        sess.run([optimizer], feed_dict={X:XData, Y:YData})
    correct_prediction = tf.equal(tf.argmax(yPred,1), tf.argmax(Y,1))
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float64))
    print(sess.run(accuracy, feed_dict={X: XDataTest, Y: YDataTest}))
    yPredn=tf.one_hot(tf.argmax(yPred,1),3)
    predn, act= sess.run([yPredn,Y], feed_dict={X: XDataTest, Y: YDataTest})
    sens=0
    count=0
    for i in range(33):
        if (i==14) | (i==24) | (i==32):
            print(sens/count)
            sens=0
            count=0
        if(np.all(predn[i]==act[i])):
            sens+=1
        count+=1
        


0.5454545454545454
0.6428571428571429
0.5
0.5


In the case of multinomial logistic regression only an accuracy of 54.54% could be achieved. Sensitivity value with respect to each outcome (<300, 300<x<450, >450) are obtained as 0.6428, 0.5, 0.5 respectively. 
Extremely low accuracy can be attributed to the very high number of features to be trained and very less training data with multiple target classes as well. Performance of model is poor with the given data.